### PR TITLE
DAOS-16365 client: intercept dlsym() and zeInit() to avoid nested calls

### PR DESCRIPTION
**Pull request by wiliamhuang**
libfabric loads libze_loader.so which calls zeInit(). We observed deadlock due to nested calls when daos_init() is called inside zeInit(). We intercept dlsym() and zeInit() to avoid calling daos_init() inside zeInit(). dlsym(RTLD_NEXT, ) checks returning address to determine caller's module. To maintain expected behavior of dlsym(RTLD_NEXT, ) with our interception, new_dlsym() is implemented with assembly code to use jmp instruction instead of call. dlsym() has been moved from libdl.so to libc.so since version 2.34.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by dinghwah — APPROVED** [review]

**Review by knard-intel — APPROVED** [review]

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 1]
```diff
@@ -1058,6 +1062,143 @@ PMPI_Init(int *argc, char ***argv)
...
+/* c code for other architecture. caller info could be wrong inside libc dlsym() when handle is set
+ * RTLD_NEXT. Assembly version implementation similar to above is needed to fix the issue by using
+ * jump instead of call instruction. 
+ */
+static void *
+new_dlsym_c(void *handle, const char *symbol)
+{
+	if (!d_hook_enabled)
+		goto org_dlsym;
+	printf("Inside my dlsym().\n");
```
Comment (tanabarr) [inline src/client/dfuse/pil4dfs/int_dfs.c:1185]: Is this debug that should be removed?

**Review by tanabarr — APPROVED** [review]
LGTM, nice work

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 2]
```diff
@@ -1058,6 +1062,143 @@ PMPI_Init(int *argc, char ***argv)
...
+			return ((void *)(addr + i + 2));
+		}
+	}
+	return NULL;
+}
+
+_Pragma("GCC diagnostic push")
+_Pragma("GCC diagnostic ignored \"-Wunused-function\"")
+_Pragma("GCC diagnostic ignored \"-Wunused-variable\"")
+
```
[the remark this answers is not in the document]
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1108]: ok. I will remove the newline here in a follow-up PR. Thank you!

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 3]
```diff
@@ -1058,6 +1062,143 @@ PMPI_Init(int *argc, char ***argv)
...
+_Pragma("GCC push_options")
+_Pragma("GCC optimize(\"-O0\")")
+static char str_zeinit[] = "zeInit";
+
+static int
+is_hook_enabled(void)
+{
+	return (d_hook_enabled ? (1) : (0));
```
[the remark this answers is not in the document]
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1116]: You are right. The inner parentheses are not necessary. I will remove them in a follow-up PR later. Thank you!

File: src/client/dfuse/pil4dfs/int_dfs.c (hunk 1 above)
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1185]: You are right. This line was planed to be removed in a follow-up PR.
This line of code is only compiled under non-x86_64 platform. It should be fine for now.

**Pull request merged**